### PR TITLE
Fix MCOL 4632 and 4648 and further issues when casting to [U]BIGINT.

### PR DESCRIPTION
**Pull request by phoeinx**

### CODE REVIEW

File: utils/funcexp/func_cast.cpp [hunk 1]
```diff
@@ -118,7 +118,8 @@ int64_t Func_cast_signed::getIntVal(Row& row, FunctionParm& parm, bool& isNull,
     case execplan::CalpontSystemCatalog::UTINYINT:
     case execplan::CalpontSystemCatalog::USMALLINT:
     {
-      return (int64_t)parm[0]->data()->getUintVal(row, isNull);
+      const auto val = static_cast<int64_t>(parm[0]->data()->getUintVal(row, isNull));
```
Comment (drrtuy) [inline utils/funcexp/func_cast.cpp]: I assume you want to apply the defined methods here?

File: utils/funcexp/func_cast.cpp [hunk 2]
```diff
@@ -143,22 +144,16 @@ int64_t Func_cast_signed::getIntVal(Row& row, FunctionParm& parm, bool& isNull,
...
-      const auto& value = parm[0]->data()->getStrVal(row, isNull);
-
-      if (isNull)
-      {
-        isNull = true;
-        return 0;
-      }
-
-      return atoll(value.str());
+      const auto& str = parm[0]->data()->getStrVal(row, isNull);
```
Comment (drrtuy) [inline utils/funcexp/func_cast.cpp]:
```suggestion
      return parm[0]->data()->getStrVal(row, isNull).toMCSInt64();
```
IMHO return looks more laconic.

File: utils/funcexp/func_cast.cpp [hunk 3]
```diff
@@ -143,22 +144,16 @@ int64_t Func_cast_signed::getIntVal(Row& row, FunctionParm& parm, bool& isNull,
...
+      return str.toMCSInt64();
     }
     break;
 
     case execplan::CalpontSystemCatalog::DECIMAL:
     case execplan::CalpontSystemCatalog::UDECIMAL:
     {
-      return parm[0]->data()->getDecimalVal(row, isNull).toSInt64Round();
+      const auto d = parm[0]->data()->getDecimalVal(row, isNull);
+      return d.toMCSInt64Round();
```
Comment (drrtuy) [inline utils/funcexp/func_cast.cpp]: ditto

File: utils/funcexp/func_cast.cpp [hunk 4]
```diff
@@ -259,24 +254,16 @@ uint64_t Func_cast_unsigned::getUintVal(Row& row, FunctionParm& parm, bool& isNu
...
-      if (isNull)
-      {
-        isNull = true;
-        return 0;
-      }
-
-      uint64_t ret = strtoul(value.str(), 0, 0);
-      return ret;
+      const auto& str = parm[0]->data()->getStrVal(row, isNull);
+      return str.toMCSUInt64();
```
Comment (drrtuy) [inline utils/funcexp/func_cast.cpp]: ditto

File: utils/funcexp/func_cast.cpp [hunk 5]
```diff
@@ -259,24 +254,16 @@ uint64_t Func_cast_unsigned::getUintVal(Row& row, FunctionParm& parm, bool& isNu
...
     }
     break;
 
     case execplan::CalpontSystemCatalog::DECIMAL:
     case execplan::CalpontSystemCatalog::UDECIMAL:
     {
-      IDB_Decimal d = parm[0]->data()->getDecimalVal(row, isNull);
-      return d.toUInt64Round();
+      const auto d = parm[0]->data()->getDecimalVal(row, isNull);
+      return d.toMCSUInt64Round();
```
Comment (drrtuy) [inline utils/funcexp/func_cast.cpp]: ditto

**Review by drrtuy — CHANGES_REQUESTED** [review]
Plz address the review comments.

File: utils/funcexp/func_cast.cpp (hunk 1 above)
Comment (phoeinx) [reply utils/funcexp/func_cast.cpp]: I didn't implement any function for converting IntValues to MCS_BIGINT values. Felt unnecessary if we can also make the logic clear here - do you I think I should also extract this logic to a method?

File: utils/funcexp/func_cast.cpp (hunk 2 above)
Comment (phoeinx) [reply utils/funcexp/func_cast.cpp]: thanks for the review - what do you mean by laconic?

Comment (drrtuy) [reply utils/funcexp/func_cast.cpp]: This means IMHO it is better to use less lines/words in the context.

File: utils/funcexp/func_cast.cpp (hunk 1 above)
Comment (drrtuy) [reply utils/funcexp/func_cast.cpp]: If you do the logic clear just here there might be another place where the same functionality is needed, namely to bound the impossible NULL value. IMHO it is better to put it into a method and use here.

File: utils/funcexp/func_cast.cpp (hunk 2 above)
Comment (phoeinx) [reply utils/funcexp/func_cast.cpp]: For `NullString` and `Decimal` the change is applicable, but for the other numeric types the return value of e.g. `getDoubleValue()` is a `long double` and not the `datatypes::TDouble` datatype on which the `toMCSSInt64Round()` function is defined. 
Therefore we would need to cast first like so:
`return ((datatypes::TDouble) parm[0]->data()->getDoubleVal(row, isNull)).toMCSSInt64Round();`
I think this is hard to read, but wanted to ask if you would still prefer it this way? I'll adapt `NullString` and `Decimal`.

Comment (phoeinx) [reply utils/funcexp/func_cast.cpp]: @drrtuy

Comment (drrtuy) [reply utils/funcexp/func_cast.cpp]: Hm, can't you use datatypes::TDouble ctor like this
`return TDouble(parm[0]->data()->getDoubleVal(row, isNull)).toMCSSInt64Round();` ?

Comment (phoeinx) [reply utils/funcexp/func_cast.cpp]: still personally think that it's a lot of text in one line, but will adapt :)

**Pull request merged**